samza-sql/src/main/java/org/apache/samza/sql/avro/AvroRelConverter.java

```diff
@@ -76,45 +80,60 @@ public AvroRelConverter(SystemStream systemStream, AvroRelSchemaProvider schemaP
    */
   @Override
   public SamzaSqlRelMessage convertToRelMessage(KV<Object, Object> samzaMessage) {
-    List<Object> fieldValues = new ArrayList<>();
-    List<String> fieldNames = new ArrayList<>();
+    List<String> payloadFieldNames = new ArrayList<>();
+    List<Object> payloadFieldValues = new ArrayList<>();
     Object value = samzaMessage.getValue();
     if (value instanceof IndexedRecord) {
-      IndexedRecord record = (IndexedRecord) value;
-      // Please note that record schema and cached schema could be different due to schema evolution.
-      // Always represent record schema in the form of cached schema. This approach has the side-effect
-      // of dropping the newly added fields in the scenarios where the record schema has newer version
-      // than the cached schema. [TODO: SAMZA-1679]
-      Schema recordSchema = record.getSchema();
-      fieldNames.addAll(avroSchema.getFields().stream()
-          .map(Schema.Field::name)
-          .collect(Collectors.toList()));
-      fieldValues.addAll(fieldNames.stream()
-          .map(f -> convertToJavaObject(
-              recordSchema.getField(f) != null ? record.get(recordSchema.getField(f).pos()) : null,
-              getNonNullUnionSchema(avroSchema.getField(f).schema())))
-          .collect(Collectors.toList()));
+      fetchFieldNamesAndValuesFromIndexedRecord((IndexedRecord) value, payloadFieldNames, payloadFieldValues,
+          payloadSchema);
     } else if (value == null) {
-      fieldNames.addAll(avroSchema.getFields().stream().map(Schema.Field::name).collect(Collectors.toList()));
-      IntStream.range(0, fieldNames.size()).forEach(x -> fieldValues.add(null));
+      payloadFieldNames.addAll(payloadSchema.getFields().stream().map(Schema.Field::name).collect(Collectors.toList()));
+      IntStream.range(0, payloadFieldNames.size()).forEach(x -> payloadFieldValues.add(null));
     } else {
       String msg = "Avro message converter doesn't support messages of type " + value.getClass();
       LOG.error(msg);
       throw new SamzaException(msg);
     }
 
-    return new SamzaSqlRelMessage(samzaMessage.getKey(), fieldNames, fieldValues);
+    Object key = samzaMessage.getKey();
+    // Key could be an avro record or could just be any other object. Handle avro here.
+    if (key instanceof IndexedRecord) {
```

Comment:
The fact that a key can be an Avro should be known to the RelMessageConverter. I don't think we need to have this if else in this layer,Rather it would be cleaner if you move this up to a layer above which uses AvroRelMessageConverter like BrooklinRelMessageConverter or KafkaRelMessageConverter.  It is ok if this layer handles keys of type String as long as there is a clean way to override the behavior.